lib/sms/sms_status_report.c

```diff
@@ -0,0 +1,386 @@
+ */
+static int parse_pdu_status_report_header(struct parser *parser, uint8_t *buf)
+{
+	struct pdu_status_report_data * const pdata = parser->data;
+
+	pdata->header = *((struct pdu_status_report_header *)buf);
+
+	LOG_DBG("SMS header 1st byte: 0x%02X", *buf);
+
+	LOG_DBG("TP-Message-Type-Indicator: %d", pdata->header.mti);
+	LOG_DBG("TP-More-Messages-to-Send: %d", pdata->header.mms);
+	LOG_DBG("TP-Loop-Prevention: %d", pdata->header.lp);
+	LOG_DBG("TP-Status-Report-Qualifier: %d", pdata->header.srq);
+	LOG_DBG("TP-User-Data-Header-Indicator: %d", pdata->header.udhi);
+
+	/* Only parse SMS-STATUS-REPORT for SMS-SUBMIT */
+	if (pdata->header.mti != TP_MTI_SMS_STATUS_REPORT ||
+	    pdata->header.srq != TP_SRQ_SMS_SUBMIT) {
+		return -EINVAL;
+	}
+
+	return 1;
+}
+
+/**
+ * @brief Decode TP-Message-Reference as specified in 3GPP TS 23.040 Section 9.2.3.6.
+ *
+ * @param[in,out] parser Parser instance.
+ * @param[in] buf Buffer containing PDU and pointing to this field.
+ *
+ * @return Number of parsed bytes.
+ */
+static int parse_pdu_mr_field(struct parser *parser, uint8_t *buf)
+{
+	struct pdu_status_report_data * const pdata = parser->data;
+
+	pdata->mr = (uint8_t)*buf;
+
+	LOG_DBG("TP-Message-Reference: %d", pdata->mr);
+
+	return 1;
+}
+
+/**
+ * @brief Decode TP-Recipient-Address as specified in 3GPP TS 23.040 Section 9.2.3.14 and 9.1.2.5.
+ *
+ * @param[in,out] parser Parser instance.
+ * @param[in] buf Buffer containing PDU and pointing to this field.
+ *
+ * @return Number of parsed bytes.
+ */
+static int parse_pdu_ra_field(struct parser *parser, uint8_t *buf)
+{
+	struct pdu_status_report_data * const pdata = parser->data;
+	uint8_t address[SMS_MAX_ADDRESS_LEN_OCTETS];
+	uint8_t length;
+
+	pdata->ra.length = *buf++;
+	pdata->ra.type = (uint8_t)*buf++;
+
+	LOG_DBG("Address-Length: %d", pdata->ra.length);
+	LOG_DBG("Type-of-Address: 0x%02X", pdata->ra.type);
+
+	if (pdata->ra.length > SMS_MAX_ADDRESS_LEN_CHARS) {
+		LOG_ERR("Maximum address length (%d) exceeded %d. Aborting decoding.",
+			SMS_MAX_ADDRESS_LEN_CHARS,
+			pdata->ra.length);
+		return -EINVAL;
+	}
+
+	length = pdata->ra.length / 2;
+
+	if (pdata->ra.length % 2 == 1) {
+		/* There is an extra number in semi-octet and fill bits*/
+		length++;
+	}
+
+	memcpy(address, buf, length);
+
+	for (int i = 0; i < length; i++) {
+		address[i] = swap_nibbles(address[i]);
+	}
+
+	convert_number_to_str(address, pdata->ra.length, pdata->ra.address_str);
+
+	/* 2 for length and type fields */
+	return 2 + length;
+}
+
+/**
+ * @brief Decode TP-Service-Centre-Time-Stamp as specified in 3GPP TS 23.040 Section 9.2.3.11.
+ *
+ * @param[in,out] parser Parser instance.
+ * @param[in] buf Buffer containing PDU and pointing to this field.
+ *
+ * @return Number of parsed bytes.
+ */
+static int parse_pdu_scts_field(struct parser *parser, uint8_t *buf)
+{
+	struct pdu_status_report_data * const pdata = parser->data;
+	int tmp_tz;
+
+	pdata->scts.year   = semioctet_to_dec(*(buf++));
```

Comment:
This code is in multiple places so we need a helper function that takes `struct sms_time*` as an output parameter and likely `uint8_t *buf` as input

Reply:
Closing this comment as SMS library changes have been removed from this PR.